### PR TITLE
added fix for WACCM ua time coordinate

### PR DESCRIPTION
**Pull request by hb326**
Time was not monotonic for a WACCM ua file, so we fixed it.
Fixing the bounds did not work. We assume because `iris.guess_bounds()` does not work for coordinate time.

### CODE REVIEW

**Comment by bouweandela** [comment]
@jvegasbsc can you please review this?

File: esmvalcore/cmor/_fixes/cmip6/cesm2_waccm.py [hunk 1]
```diff
@@ -1,6 +1,42 @@
...
+            return cubes
+        coord = cube.coord('time')
+        idx_sorted = np.argsort(coord.points)
+        coord.points = coord.points[idx_sorted]
+        coord.bounds = None
+        # coord.guess_bounds('time')  # fails for coordinate 'time'
+        time_idx = cube.coord_dims('time')[0]
+        slice_idx = [slice(None)] * cube.ndim
+        slice_idx[time_idx] = idx_sorted
+        cube.data = cube.data[slice_idx]
```
Comment (bouweandela) [inline esmvalcore/cmor/_fixes/cmip6/cesm2_waccm.py]:
```suggestion
        cube.data = cube.core_data()[slice_idx]
```
this avoid loading all data into memory

File: tests/integration/cmor/_fixes/cmip6/test_cesm2_waccm.py [hunk 2]
```diff
@@ -39,3 +50,22 @@ def test_tas_fix_metadata(tas_cubes):
...
                 cube.coord('height')
+
+
+def test_get_ua_fix():
+    fix = Fix.get_fixes('CMIP6', 'CESM2-WACCM', 'ua')
+    assert fix == [Ua()]
+
+
+def test_ua_fix_metadata(ua_cubes):
+    expected_coord = iris.coords.AuxCoord(
```
Comment (bouweandela) [inline tests/integration/cmor/_fixes/cmip6/test_cesm2_waccm.py]:
```suggestion
    expected_coord = iris.coords.DimCoord(
```

File: tests/integration/cmor/_fixes/cmip6/test_cesm2_waccm.py [hunk 3]
```diff
@@ -39,3 +50,22 @@ def test_tas_fix_metadata(tas_cubes):
...
+        [1.0, 2.0, 3.0], standard_name='time', units='days since 1850-01-01',
+        bounds=None)
+    expected_cube = iris.cube.Cube([7.0, 9.0, 8.0], var_name='ua',
+                                   aux_coords_and_dims=[(expected_coord, 0)])
```
Comment (bouweandela) [inline tests/integration/cmor/_fixes/cmip6/test_cesm2_waccm.py]:
```suggestion
                                   dim_coords_and_dims=[(expected_coord, 0)])
```

File: esmvalcore/cmor/_fixes/cmip6/cesm2_waccm.py [hunk 4]
```diff
@@ -1,6 +1,42 @@
...
+        """
+        cube = self.get_cube_from_list(cubes)
+        dim_coords = [coord.name() for coord in cube.coords(dim_coords=True)]
+        if 'time' in dim_coords:
+            return cubes
+        coord = cube.coord('time')
+        idx_sorted = np.argsort(coord.points)
+        coord.points = coord.points[idx_sorted]
+        coord.bounds = None
+        # coord.guess_bounds('time')  # fails for coordinate 'time'
```
Comment (bouweandela) [inline esmvalcore/cmor/_fixes/cmip6/cesm2_waccm.py]: What is failing exactly?

**Comment by JaroCamphuijsen** [comment]
@hb326 can you have a look at the comments above? Then we can probably merge this.

**Comment by valeriupredoi** [comment]
OK I started committing stuff that @bouweandela suggested (I know, I'm nasty - changing code in a PR that has nothing to do with mesefl heh, but was trying to see if we can get this through to the 2.2 release) but I stopped coz it looks like we really need @hb326 to have a look and comment/commit :+1:

**Comment by hb326** [comment]
> 
> 
> OK I started committing stuff that @bouweandela suggested (I know, I'm nasty - changing code in a PR that has nothing to do with meself heh, but was trying to see if we can get this through to the 2.2 release) but I stopped coz it looks like we really need @hb326 to have a look and comment/commit 👍

I am looking at it now. Sorry, took a while. Hopefully we can get it sorted for release 2.3!

**Comment by valeriupredoi** [comment]
cool stuff! just ping me here if you need a review :beer:

File: esmvalcore/cmor/_fixes/cmip6/cesm2_waccm.py (hunk 4 above)
Comment (zklaus) [reply esmvalcore/cmor/_fixes/cmip6/cesm2_waccm.py:36]: Agree with @bouweandela question. @hb326, note that `coord` already refers to time, so I am not sure why there is another mention of time in the args. Certainly `guess_bounds` works with time coordinates in general.

**Comment by hb326** [comment]
@valeriupredoi: it seems like the problem vanished! I just now ran a recipe where I try to read the WACCM data without any fix, in different variations (historical, amip, piControl), and it seems like everything is read correctly. Could you please check if you can read the following data, no need for a script, just reading the data with the ESMValTool:
- {dataset: CESM2-WACCM,  project: CMIP6,  mip: day,  exp: historical,  grid: gn, ensemble: r1i1p1f1,  start_year: 1960,  end_year: 2014} 
Maybe some update to the Core has fixed the problem without the dataset fix?

**Comment by hb326** [comment]
> 
> 
> @valeriupredoi: it seems like the problem vanished! I just now ran a recipe where I try to read the WACCM data without any fix, in different variations (historical, amip, piControl), and it seems like everything is read correctly. Could you please check if you can read the following data, no need for a script, just reading the data with the ESMValTool:
> 
>     * {dataset: CESM2-WACCM,  project: CMIP6,  mip: day,  exp: historical,  grid: gn, ensemble: r1i1p1f1,  start_year: 1960,  end_year: 2014}
> 
> 
> Maybe some update to the Core has fixed the problem without the dataset fix?

I should add that I am using the current master branch for the ESMValTool and the ESMValCore for this test.

**Comment by hb326** [comment]
@valeriupredoi: did you check yet if WACCM would run for you without fix as well?

**Review by valeriupredoi — CHANGES_REQUESTED** [review]
@hb326 I understand the special case but I think it's best you add a note (in-line comment) about this particular case (it was visible to me from the test case, cheers for adding it!). But my question is - what sort of DimCoord does the cube have to account for the time axis, if time itself is an AuxCoord

**Comment by valeriupredoi** [comment]
@hb326 could you pls do a mergy merge of `master` into this, some bits are causing age-related issues, and fix the conflicts as well pls (lemme know if you need help with that) :beer:

**Comment by hb326** [comment]
> 
> 
> @hb326 could you pls do a mergy merge of `master` into this, some bits are causing age-related issues, and fix the conflicts as well pls (lemme know if you need help with that) 🍺

@valeriupredoi, my thinking is that we can just forget this pull request since it seems to me that some change in the ESMValCore actually fixed the problem for which I had written the fix. Did you try to read in the WACCM ua data with the current master branch of ESMValCore and ESMValTool? Because for me it works like a charm, WITHOUT using the fix.

**Comment by hb326** [comment]
Ok, it does seem like the problem that this PR was supposed to fix was fixed with an update to the ESMValCore recently. So this PR is redundant.

**Comment by valeriupredoi** [comment]
fairly sure it's the data that got updated @hb326

**Pull request closed without merging**